engine/src/main/java/org/hibernate/validator/internal/metadata/provider/AnnotationMetaDataProvider.java

```diff
@@ -791,62 +776,62 @@ else if ( annotatedType instanceof AnnotatedParameterizedType ) {
 		return MetaConstraints.create( typeResolutionHelper, valueExtractorManager, descriptor, constraintLocation );
 	}
 
-	private CascadingMetaDataBuilder getCascadingMetaData(Type type, AnnotatedElement annotatedElement,
+	private CascadingMetaDataBuilder getCascadingMetaData(JavaBeanAnnotatedElement annotatedElement,
 			Map<TypeVariable<?>, CascadingMetaDataBuilder> containerElementTypesCascadingMetaData) {
-		return CascadingMetaDataBuilder.annotatedObject( type, annotatedElement.isAnnotationPresent( Valid.class ), containerElementTypesCascadingMetaData,
-						getGroupConversions( annotatedElement ) );
+		return CascadingMetaDataBuilder.annotatedObject( annotatedElement.getType(), annotatedElement.isAnnotationPresent( Valid.class ),
+				containerElementTypesCascadingMetaData, getGroupConversions( annotatedElement.getAnnotatedType() ) );
 	}
 
 	/**
 	 * The location of a type argument before it is really considered a constraint location.
 	 * <p>
 	 * It avoids initializing a constraint location if we did not find any constraints. This is especially useful in
-	 * a Java 9 environment as {@link ConstraintLocation#forProperty(Member) tries to make the {@code Member} accessible
+	 * a Java 9 environment as {@link ConstraintLocation#forProperty(Property)} tries to make the {@code Member} accessible
```

Comment:
This is actually not true anymore, as now we are making things accessible right away in `JavaBeanField` or `JavaBeanGetter` constructors. Hence two questions:
- do we need these `TypeArgumentLocation`s ?
- do we need to make things accessible lazily in those `JavaBean*` constructors?

Reply:
Hum. That is indeed an interesting point. I missed that in the various refactorings.

This is annoying :/.

> do we need these ``TypeArgumentLocation``s ?

If you find a clean way to remove them, we can get rid of them. But let's do it in a follow-up PR.

> do we need to make things accessible lazily in those JavaBean* constructors?

Yes, we do. We shouldn't make the field accessible until we are sure there are constraints (or cascading) and we will need the value.

I'm not a big fan of using a double checked locking mechanism as it will be in a hot path so I tried to introduce a `PropertyAccessor` in a follow-up commit.

Another alternative would be to make `JavaBeanField` and `JavaBeanGetter` builders but I'm not sure it's going to be better.

A drawback of this approach is that we create 2 instances of the accessible fields. But we already did that.

WDYT?